zeppelin-server/src/main/java/org/apache/zeppelin/service/InterpreterService.java

```diff
@@ -0,0 +1,155 @@
+
+  private static final Logger logger = LoggerFactory.getLogger(InterpreterService.class);
+  private static final ExecutorService executorService =
+      Executors.newSingleThreadExecutor(
+          new ThreadFactoryBuilder()
+              .setNameFormat(InterpreterService.class.getSimpleName() + "-")
+              .build());
+
+  private final ZeppelinConfiguration conf;
+  private final InterpreterSettingManager interpreterSettingManager;
+
+  public InterpreterService(
+      ZeppelinConfiguration conf,
+      InterpreterSettingManager interpreterSettingManager) {
+    this.conf = conf;
+    this.interpreterSettingManager = interpreterSettingManager;
+  }
+
+  public void installInterpreter(
+      final InterpreterInstallationRequest request,
+      final MessageCallback startedMessage,
+      final MessageCallback finishedMessage)
+      throws Exception {
+    Preconditions.checkNotNull(request);
+    Preconditions.checkNotNull(request.getName());
+    Preconditions.checkNotNull(request.getArtifact());
+
+    String interpreterBaseDir = conf.getInterpreterDir();
+    String localRepoPath = conf.getInterpreterLocalRepoPath();
+
+    final DependencyResolver dependencyResolver = new DependencyResolver(localRepoPath);
+
+    String proxyUrl = conf.getZeppelinProxyUrl();
+    if (null != proxyUrl) {
+      String proxyUser = conf.getZeppelinProxyUser();
+      String proxyPassword = conf.getZeppelinProxyPassword();
+      try {
+        dependencyResolver.setProxy(new URL(proxyUrl), proxyUser, proxyPassword);
+      } catch (MalformedURLException e) {
+        // TODO(jl): Not sure if it's good to raise an exception
+        throw new Exception("Url is not valid format", e);
+      }
+    }
+
+    final Path interpreterDir = Paths.get(interpreterBaseDir, request.getName());
+    if (Files.exists(interpreterDir)) {
+      throw new Exception("Interpreter " + request.getName() + " already exists");
+    }
+
+    try {
+      Files.createDirectories(interpreterDir);
+    } catch (Exception e) {
+      throw new Exception("Cannot create " + interpreterDir.toString());
+    }
+
+    // It might take time to finish it
+    executorService.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            downloadInterpreter(
+                request, dependencyResolver, interpreterDir, startedMessage, finishedMessage);
+          }
+        });
+  }
+
+  void downloadInterpreter(
+      InterpreterInstallationRequest request,
+      DependencyResolver dependencyResolver,
+      Path interpreterDir,
+      MessageCallback startedMessage,
+      MessageCallback finishedMessage) {
+    Message m = new Message(OP.INTERPRETER_INSTALL_STARTED);
+    Map<String, Object> result = Maps.newHashMap();
+    try {
+      logger.info("Start to download a dependency: {}", request.getName());
+      result.put("result", "Starting");
+      result.put("message", "Starting to download " + request.getName() + " interpreter");
```

Comment:
Thanks for introducing Callback, but I think we can do more.
1. How about put them in one callback class.
```
  interface MessageCallback {
     onStart();
     onSuccess();
     onFailure();
  }
```

2. All the message stuff should be put in the callback instead of putting them in the Service class.  E.g. 
Here we can constructing the result HashMap in callback interface. The purpose is that Service class just do the service work, and don't need to care about what do do when service is succeeded or failure, it is callback's responseentity to that do.

Reply:
Agreed. Fixed